File: day10/solve.py
from runner.python import solve


def parse_map(input: str) -> list[list[int]]:
  return [list(map(int, line)) for line in input.split("\n")]


def find_trailheads(map: list[list[int]]) -> list[tuple[int, int]]:
  return [
    (x, y)
    for y in range(len(map))
    for x in range(len(map[0]))
    if map[y][x] == 0
  ]


def find_end_of_all_trails(
  x: int,
  y: int,
  map: list[list[int]],
) -> list[tuple[int, int]]:
  if map[y][x] == 9:
    return [(x, y)]

  return [
    trail_end
    for next_x, next_y in [(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)]
    if 0 <= next_y < len(map)
    and 0 <= next_x < len(map[0])
    and map[next_y][next_x] == map[y][x] + 1
    for trail_end in find_end_of_all_trails(next_x, next_y, map)
  ]


def part1(input: str) -> int:
  map = parse_map(input)
  return sum(
    len(set(find_end_of_all_trails(x, y, map)))
    for x, y in find_trailheads(map)
  )


def part2(input: str) -> int:
  map = parse_map(input)
  return sum(
    len(find_end_of_all_trails(x, y, map)) for x, y in find_trailheads(map)
  )


solve(
  part1,
  "89010123\n78121874\n87430965\n96549874\n45678903\n32019012\n01329801\n10456732",
  36,
  part2,
  "89010123\n78121874\n87430965\n96549874\n45678903\n32019012\n01329801\n10456732",
  81,
)

File: day10/solve.ts
import { solve } from "../runner/typescript";
import { max, sum, groupBy, uniq } from "lodash";

function parser(input: string) {
  return input.split("\n").map((l) => l.split("").map(Number));
}

function findEndOfAllTrails(
  map: number[][],
  x: number,
  y: number
): `${number},${number}`[] {
  const current = map[y][x];
  if (current === 9) return [`${x},${y}`]; // TODO what if already found?

  return [
    [x - 1, y],
    [x + 1, y],
    [x, y - 1],
    [x, y + 1],
  ]
    .filter(
      ([nextX, nextY]) =>
        !!map[nextY]?.[nextX] && map[nextY][nextX] === current + 1
    )
    .flatMap(([nextX, nextY]) => findEndOfAllTrails(map, nextX, nextY));
}

function part1(map: number[][]): number {
  // coordinates of all 0s on the map
  const trailheads = map.flatMap((row, y) =>
    row.flatMap((value, x) => (value === 0 ? [[x, y]] : []))
  );

  console.log({ trailheads });
  const scores = trailheads.map(
    ([x, y]) => uniq(findEndOfAllTrails(map, x, y)).length
  );

  console.log({ scores });
  return sum(scores);
}

function part2(map: number[][]): number {
  return 0;
}

solve({
  parser,
  part1,
  // part2,

  part1Tests: [
    ["...0...\n...1...\n...2...\n6543456\n7.....7\n8.....8\n9.....9", 2],
    [
      "89010123\n78121874\n87430965\n96549874\n45678903\n32019012\n01329801\n10456732",
      36,
    ],
  ],
  part2Tests: [
    // ["aaa", 0],
    // ["a", 0],
  ],
});
